libres/tests/res_util/test_matrix.cpp

```diff
@@ -68,3 +69,28 @@ TEST_CASE("matrix_subtract_row_mean", "[res_util]") {
 
     matrix_free(X);
 }
+
+SCENARIO("matrix_dgesvx") {
+    GIVEN("A diagonal matrix A") {
+        std::size_t N = 4;
+        matrix_type *A = matrix_alloc(N, N);
+        for (int i = 0; i < N; i++)
+            matrix_iset(A, i, i, 1.0 / (i + 1));
+
+        matrix_type *bx = matrix_alloc(N, 1);
+        for (int i = 0; i < N; i++)
+            matrix_iset(bx, i, 0, 1);
+
+        WHEN("dgesvx is called with cond argument") {
+            double rcond;
+            matrix_dgesvx(A, bx, &rcond);
+        }
+
+        WHEN("dgesvx is called with nullptr cond argument") {
+            matrix_dgesvx(A, bx, nullptr);
+        }
+
+        for (int i = 0; i < N; i++)
```

Comment:
For readability of the output in catch, it would be helpful to have a `THEN` block around this loop describing the expected behavior.

Reply:
Well - if the test is reorganized as 
```
WHEN() { ... }

WHEN() { ... }

THEN() { ... }
```
it will fail - because only one section at a time is walked all the way to the leaf nodes, and with `THEN` in a separate section we will get two setup-and-calculate sections and one separate check section.

Reply:
That is true, but you can use `GENERATE` to drop one of your `WHEN` blocks. Personally I would use:
```
SCENARIO("Solving a linear equation with dgesvx") {
    GIVEN("A diagonal matrix A") {
        std::size_t N = 4;
        matrix_type *A = matrix_alloc(N, N);
        for (int i = 0; i < N; i++)
            matrix_iset(A, i, i, 1.0 / (i + 1));

        AND_GIVEN("A vector B with all entries equal to 1"){
            matrix_type *bx = matrix_alloc(N, 1);
            for (int i = 0; i < N; i++)
                matrix_iset(bx, i, 0, 1);

            WHEN("Using dgesvx to solve the linear equation Ax=B") {
                double rcond;
                matrix_dgesvx(A, bx, GENERAGE(&rcond, nullptr));

                THEN("B is modified to be the solution") {
                    for (int i = 0; i < N; i++)
                        REQUIRE(matrix_iget(bx, i, 0) == (i + 1));
                }
            }
        }
    }
}
```

Reply:
That is true, but you can avoid needing multiple `WHEN` blocks with a `GENERATE`. Personally I would use:
```

SCENARIO("Solving a linear equation with dgesvx") {
    GIVEN("A diagonal matrix A") {
        std::size_t N = 4;
        matrix_type *A = matrix_alloc(N, N);
        for (int i = 0; i < N; i++)
            matrix_iset(A, i, i, 1.0 / (i + 1));

        AND_GIVEN("A vector B with all entries equal to 1"){
            matrix_type *bx = matrix_alloc(N, 1);
            for (int i = 0; i < N; i++)
                matrix_iset(bx, i, 0, 1);

            WHEN("Using dgesvx to solve the linear equation Ax=B") {
                double * rcond = GENERATE((double*)malloc(sizeof(double)), NULL);
                matrix_dgesvx(A, bx, rcond);

                THEN("B is modified to be the solution") {
                    for (int i = 0; i < N; i++)
                        REQUIRE(matrix_iget(bx, i, 0) == (i + 1));
                }
                free(rcond);
            }
        }
    }
}
```

Reply:
Well - this is different test; it only runs (and verifies) the matrix solve for `rcond != nullptr`.

Reply:
I think it will run with `rcond=NULL` as well because of `GENERATE` :)

Reply:
Ok - thanks; I had not understood GENERATE

Reply:
OK - updated; valuable to learn about GENERATE.